### PR TITLE
Add video classification pipeline

### PR DESCRIPTION
**Pull request by nateraw**
# What does this PR do?

Adds a video classification pipeline using VideoMAE.

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/pipelines/video_classification.py
Comment (nateraw) [inline src/transformers/pipelines/video_classification.py:37]: Need this information but its not in model config or preprocessor. Whats best place for this to be?

**Review by nateraw — COMMENTED** [review]
Some thoughts

File: src/transformers/pipelines/video_classification.py
Comment (amyeroberts) [inline src/transformers/pipelines/video_classification.py:89]: Do we need `np.clip` here?

File: tests/pipelines/test_pipelines_video_classification.py
Comment (amyeroberts) [reply tests/pipelines/test_pipelines_video_classification.py:65]: nit: the `size` param should also be a dict
```suggestion
        small_feature_extractor = VideoMAEFeatureExtractor(size=dict(shortest_edge=10), crop_size=dict(height=10, width=10))
```

**Review by amyeroberts — APPROVED** [review]
LGTM! Thanks for adding ❤️

File: src/transformers/pipelines/video_classification.py
Comment (nateraw) [reply src/transformers/pipelines/video_classification.py:89]: Ah I changed the code a bit to do the `end_idx - 1` bit earlier, so the clip is probably unnecessary now. good catch...will double check and remove

File: src/transformers/pipelines/video_classification.py
Comment (alaradirik) [inline src/transformers/pipelines/video_classification.py:79]:
```suggestion
```

**Review by alaradirik — APPROVED** [review]
Thanks for adding this, LGTM!

**Review by sgugger — APPROVED** [review]
Thanks for adding this! Also pinging @Narsil for review as it's a new pipeline.

File: src/transformers/pipelines/video_classification.py
Comment (Narsil) [inline src/transformers/pipelines/video_classification.py:41]: We should use `_sanitize_parameters` for `frame_sampling_rate`, we cannot rely on `self` for parameters because we always need to be able to know which values to use.

```python
pipe = pipeline(..., arg=1)
out = pipe(..) # Using arg=1
out = pipe(arg=2) # Using arg=2
out = pipe(...) # Using arg=1 again.
```

This is the reason for this weird `_sanitize_parameters` thing. To have a single location that handles that complexity.

`self.num_frames` seems to be used only in `preprocess` and should be defined only there. (without self).
Keep in mind that `preprocess` might be working on a different thread than main thread.

File: src/transformers/pipelines/video_classification.py
Comment (Narsil) [inline src/transformers/pipelines/video_classification.py:83]: What is the `VideoReader` object ?

I don't feel good about using stuff like threading of `cpu(0)` as `preprocess` is already supposed to be on specific threads (with `DataLoader` for PyTorch). 

If there is a simple single threaded load variant I think it would fit better.

If video loading uses all threads by default and it makes more sense that way we might want to handle the case where the pipeline is suposed to start multithreading to deactivate it.

Comment (Narsil) [reply src/transformers/pipelines/video_classification.py:83]: Actually went to check the dependency and I'm not sure it's worthwile, if it's using `ffmpeg` under the hood, we can write a simple function that will do what's required using subprocess quite easily.

File: src/transformers/pipelines/video_classification.py
Comment (Narsil) [inline src/transformers/pipelines/video_classification.py:92]: Why casting to `list` ? We cannot work with a real tensor ?

Shouldn't the tensor already be in `torch` format ?

File: tests/pipelines/test_pipelines_video_classification.py
Comment (Narsil) [inline tests/pipelines/test_pipelines_video_classification.py:94]: Do we have a large model to make sure the test actually works as intended and classifies this video as archery ?

**Review by Narsil — COMMENTED** [review]
Overall looks super clean !

We need to modify the parameter handling as I suggested.

I have doubts about the necessity of using `decord` for simple video loading.
If performance is in question, I'm happy to take a stab at this.

File: src/transformers/pipelines/video_classification.py
Comment (nateraw) [reply src/transformers/pipelines/video_classification.py:83]: The problem with ffmpeg + subprocess is that it will fail silently, no?

File: src/transformers/pipelines/video_classification.py
Comment (nateraw) [reply src/transformers/pipelines/video_classification.py:92]: This is library agnostic...its a list of numpy frames. The video classification feature extractor expects this as a list, not numpy array, which is why I did it this way

File: src/transformers/pipelines/video_classification.py
Comment (Narsil) [reply src/transformers/pipelines/video_classification.py:83]: Depends on the code, mostly no. At least it shouldn't.

File: src/transformers/pipelines/video_classification.py
Comment (Narsil) [reply src/transformers/pipelines/video_classification.py:92]: Casting objects around creates many copies is usually a good way to slow things down unecessarily.
For video it's especially important to be as efficient as possible IMO. (Like copying small strings or 1-d token id tensor is usually not that impactful, but entire decoded videos much more so)

File: tests/pipelines/test_pipelines_video_classification.py
Comment (nateraw) [reply tests/pipelines/test_pipelines_video_classification.py:94]: The default model [MCG-NJU/videomae-base-finetuned-kinetics](https://huggingface.co/MCG-NJU/videomae-base-finetuned-kinetics) works for this.

File: src/transformers/pipelines/video_classification.py
Comment (nateraw) [reply src/transformers/pipelines/video_classification.py:92]: Totally agree. Unfortunately, it seems the `feature_extractor` here is looking for a list of np arrays or PIL images. Passing the whole video array and processing it all at once would be much better, but it seems that's not how its implemented?

So this issue isn't really with this PR it's rooted in the processor I think. WDYT I should do?

Other option with current implem of `feature_extractor` is to do something like `video = [x.asnumpy() for x in videoreader.get_batch(indices)]` but I assume that's slower than the way I did it here.

Comment (Narsil) [reply src/transformers/pipelines/video_classification.py:92]: Could be left as another PR, but I think the feature extractor should be able to receive the array as-is, I don't see any valid reasons to be forced to receive list of PIL image (although it's nice to be able to receive that).

I looked at the feature extractor, and it does center cropping and some normalization only, it feels to me that we could do all that at load time of the video meaning it should be mostly a no-op (the normalize should be simple add on the tensor iiuc).

File: tests/pipelines/test_pipelines_video_classification.py
Comment (Narsil) [reply tests/pipelines/test_pipelines_video_classification.py:94]: Perfect let's add a slow test then.

**Comment by nateraw** [comment]
Holding off on this PR as we discuss #5225 - I think I will update the PR here to use `av` instead of `decord` because of it. Feel free to join the conversation there.

**Comment by Narsil** [comment]
> Holding off on this PR as we discuss #5225 - I think I will update the PR here to use `av` instead of `decord` because of it. Feel free to join the conversation there.

Your link is wrong I think, you meant https://github.com/huggingface/datasets/issues/5225 (Tip: GH will shorten the URL on its own so you don't have to care, just copy&paste raw URLs :) )

Maybe a core maintainer could jump in, but I feel like "blocking" PRs like this is not desirable, we should merge whatever is ready first, and hardmonize later. if this PRs code isolate the dependency enough, it should be  a breeze to update.
And if it's not it could be an argument in favor/defavor of some library. Real code always trumps whatever feelings about library X.

**Comment by sgugger** [comment]
I agree the PR should not be held off until a feature is merged in Datasets. We can adapt to it later on when Datasets has the features.

**Comment by nateraw** [comment]
Ok thanks for the advice @sgugger - in that case I'll just resolve all PR comments here and finish this out this week.

File: tests/pipelines/test_pipelines_video_classification.py
Comment (nateraw) [reply tests/pipelines/test_pipelines_video_classification.py:96]: I believe the `run_pipeline_test` is doing this already? not sure. Looks like its grabbing the default model and doing this already

**Comment by nateraw** [comment]
@Narsil is it ok to leave decord for now? I think its fine for this use case, and is just constrained to this pipeline. Later, we'll probably want to add some `video_utils.py` file, just as we do with image utils, where we can keep some more permanent video utilities.

Based on the convo in the datasets repo, I think we'll end up using PyAV.

To try this feature:

```python
from transformers import pipeline

pipe = pipeline('video-classification')
pipe('https://huggingface.co/datasets/nateraw/video-demo/resolve/main/archery.mp4')

# Result
"""
[{'score': 0.6418354511260986, 'label': 'archery'}, {'score': 0.0026529659517109394, 'label': 'riding unicycle'}, {'score': 0.00258301617577672, 'label': 'golf driving'}, {'score': 0.002545431721955538, 'label': 'throwing ball'}, {'score': 0.0023797585163265467, 'label': 'tobogganing'}]
"""
```

**Pull request merged**
